jbpm-wb-case-mgmt/jbpm-wb-case-mgmt-client/src/main/java/org/jbpm/workbench/cm/client/roles/CaseRolesPresenter.java

```diff
@@ -48,120 +58,164 @@ public String getTittle() {
+                                editRoleAssignmentView.getGroupsNames()
+                        ));
+            }
+        };
     }
 
-    protected void addAssignment(final String userName,final String groupName, final String roleName) {
-        if(!isNullOrEmpty(userName) && !isNullOrEmpty(groupName)){
-            caseService.call(
-                    (Void) -> findCaseInstance()
-            ).assignGroupAndUserToRole(serverTemplateId, containerId, caseId, roleName, userName,groupName);
-        }else {
-            if (!isNullOrEmpty(userName)) {
-                addUserToRole(userName, roleName);
+    private CaseRoleAction getRemoveAction(String roleName, boolean enabledAction) {
+        return new CaseRoleAction() {
+            @Override
+            public String label() {
+                return translationService.format(REMOVE_ALL_ASSIGNMENTS);
+            }
+
+            @Override
+            public boolean isEnabled() {
+                return enabledAction;
             }
-            if(!isNullOrEmpty(groupName)){
-                addGroupToRole(groupName,roleName);
+
+            @Override
+            public void execute() {
+                confirmPopup.show(translationService.format(REMOVE_ASSIGNMENT),
+                        translationService.format(REMOVE),
+                        translationService.format(REMOVE_ALL_USERS_GROUPS_FROM_ROLE, roleName),
+                        () -> assignToRole(roleName, Collections.emptyList(),
+                                Collections.emptyList()));
             }
+        };
+    }
+
+    private boolean isAssignmentAvailable(CaseDefinitionSummary cds, CaseInstanceSummary cis, String roleName) {
+        final Integer roleCardinality = cds.getRoles().get(roleName);
+        if (roleCardinality == -1) {
+            return true;
         }
+        final Integer roleInstanceCardinality =
+                cis.getRoleAssignments().stream()
+                        .filter(ra -> roleName.equals(ra.getName()))
+                        .findFirst()
+                        .map(ra -> ra.getGroups().size() + ra.getUsers().size()).orElse(0);
+        return roleInstanceCardinality < roleCardinality;
     }
 
-    protected void addUserToRole(final String userName, final String roleName) {
-        caseService.call(
-                (Void) -> findCaseInstance()
-        ).assignUserToRole(serverTemplateId, containerId, caseId, roleName, userName);
+    protected CaseRoleAssignmentSummary getRoleAssignment(final CaseInstanceSummary cis, final String roleName) {
+        if (cis.getRoleAssignments() != null && cis.getRoleAssignments().size() > 0) {
+            return cis.getRoleAssignments()
+                    .stream()
+                    .filter(ra -> roleName.equals(ra.getName()))
+                    .findFirst().orElse(CaseRoleAssignmentSummary.builder().name(roleName).build());
+        }
+        return CaseRoleAssignmentSummary.builder().name(roleName).build();
+
+    }
+
```

Comment:
Just a suggestion how to avoid returning "dummy" CaseRoleAssignmentSummary instances which doesn't seem to serve any purpose where this method is called:
```java
protected Optional<CaseRoleAssignmentSummary> getRoleAssignment(final CaseInstanceSummary cis, final String roleName) {
        if (cis.getRoleAssignments() != null && cis.getRoleAssignments().size() > 0) {
            return cis.getRoleAssignments()
                    .stream()
                    .filter(ra -> roleName.equals(ra.getName()))
                    .findFirst();
        }
        return Optional.empty();
```

Then in setupExistingAssignments method you are only taking into account CaseRoleAssignmentSummary instances which have the role with the name you are searching:
```java
cds.getRoles().keySet().stream().forEach(
                            roleName -> {
                                Optional<CaseRoleAssignmentSummary> caseRoleAssignmentSummary = getRoleAssignment(cis, roleName);
                                if(caseRoleAssignmentSummary.isPresent()) {
                                    List<CaseAssignmentItem> usersList = new ArrayList<CaseAssignmentItem>();
                                    caseRoleAssignmentSummary.get().getUsers()
                                            .forEach(u -> usersList.add(new CaseAssignmentItem() {
                                                @Override
                                                public String label() {
                                                    return u;
                                                }
...
```